domain/application/service/application.go

```diff
@@ -18,14 +18,24 @@ import (
 	"github.com/juju/juju/domain/application"
 	domaincharm "github.com/juju/juju/domain/application/charm"
 	applicationerrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/domain/life"
 	domainstorage "github.com/juju/juju/domain/storage"
+	uniterrors "github.com/juju/juju/domain/unit/errors"
 	"github.com/juju/juju/internal/charm"
 	"github.com/juju/juju/internal/storage"
 )
 
 // ApplicationState describes retrieval and persistence methods for
 // applications.
 type ApplicationState interface {
+	// ExecuteTxnOperation starts a txn and looks up ID of the specified application.
+	// It invokes the supplied callback with the app ID and a set of operations which
+	// can be called and will run in the single transaction. It returns an error
+	// satisfying [applicationerrors.ApplicationNotFoundError] if the application doesn't exist
+	// and an error satisfying [applicationerrors.ApplicationIsDead] if the application is dead
+	// and deadOk is false.
+	ExecuteTxnOperation(ctx context.Context, appName string, deadOk bool, _ application.StateTxOperationFunc) error
```

Comment:
@wallyworld I meant to discuss this with you before/after the spec review meet, but was unable to.

I don't believe we should do this. It tends towards what we have today in Mongo.

As implemented, there would be a tendency for `StateTxOperation` interfaces to become "god interfaces" containing all state methods required by transactions with business logic. The alternative is not really better, as there would be a proliferation of such interfaces, each requiring their own version of `StateTxOperationFunc`.

This would make the domains increasingly opaque over time and place cognitive burden on (especially more junior) engineers.

I am increasingly of the opinion that what we have at present is actually fine - **just push business-based branching into the state layer transactions as required**. This is the clearest representation of logic available to us.

It begs the question though: _what is the point of service/state separation?_ We did this to separate business logic from the persistence layer.

If we hypothesise and imagine the other of the scenarios we discussed at the (can't remember which) engineering sprint - just having a state layer, what would be the implications? 
- Validation intermingled with transactional logic.
- Provider interaction intermingled with transactional logic.
- Watcher acquisition intermingled with transactional logic.

The separation does ensure one thing, even if we make business logic decisions inside transactions. We can't shoot ourselves in the foot by holding open transactions for any of the operations listed above. The state layer remains only concerned with reading, interacting with, and writing data.

Just write the transactions that we need as we've done for the secrets domain.

Reply:
This makes me very sad that we did not get the txn boundaries done at the right layer and so are using that as a driver to again intermingle service business logic with persistence logic. We did that with mongo and it's been a bad experience. We are substituting mongo asserts etc for sql queries. We must IMO be very clear that the state layer is purely for transforming an in memory domain model to a persistent representation. We must not blur the separation of concerns between algorithmic business logic and persistence.

The secrets domain is a case where we have to date punted on implementing service logic not in a single txn, pending the outcome of this discussion. We have avoided pushing service logic into the state layer because it simply does not belong there.

Reply:
In terms of the structure of it - the service business logic still uses bespoke, fit for purpose state api calls to read and write domain data; the state methods are designed to fit the use case / workflow at hand. The service logic could be copied directly from the closure and run directly in the service method without the closure if the state methods on the txn operations interface were copied to the state interface. All we are doing here is affording the service logic the opportunity to run exactly the same code as it would have run but in the context of a single transaction. The separation of concerns is maintained and we avoid mixing service business logic with sql which would be a Disaster V2.0 in X years time.

Reply:
> The service logic could be copied directly from the closure and run directly in the service method without the closure if the state methods on the txn operations interface were copied to the state interface.

You reintroduce the possibility of non-persistence concerns in transactions, such as going out to the provider.

> All we are doing here is affording the service logic the opportunity to run exactly the same code as it would have run but in the context of a single transaction.

Just run it in the transaction. It's clearer.

Reply:
I'd like to start by summarizing my interpretation of both points of view. It will be a bit simplified, but hopefully makes sure that I'm actually understanding things correctly.

Ian would like to see most API methods be done in a single transaction, as being able to 'read your inputs and define your output' is a very nice level. As it also makes it much clearer what would need to be retried (you need to reassess all your inputs in the case that the change doesn't work, and you should have a single transaction so that you know your inputs haven't changed by the time you write your output).

Joe is concerned that there are some inputs that can't be done strictly transactionally because they aren't database level interactions. Reading from a Provider, interacting with Watchers, etc, are potentially blocking operations that would be holding a DB transaction for arbitrary lengths of time. An additional concern is that if we have to break up the logic into closures that get passed, we end up making the actual interactions harder to reason about.

I still need to dig into some examples to see the value/downsides of both approaches, but I at least wanted to start the discussion.